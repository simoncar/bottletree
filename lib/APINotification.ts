import {
  addDoc,
  doc,
  updateDoc,
  collection,
  onSnapshot,
  setDoc,
  Timestamp,
  getDocs,
  documentId,
  orderBy,
} from "firebase/firestore";
import { auth, db } from "./firebase";
import { IPushToken, IUser } from "./types";
import * as Notifications from "expo-notifications";

// ExponentPushToken[z-50OyGeRPth6nxZSWk_A4]

export function setToken(
  token: IPushToken,
  callback: { (id: string): void; (arg0: string): void },
) {
  try {
    const docRef = doc(db, "tokens", token.key);

    console.log(
      "setToken",
      token.key,
      token.pushToken,
      auth.currentUser?.uid,
      auth.currentUser?.displayName,
    );

    setDoc(
      docRef,
      {
        pushToken: token.pushToken,
        uid: auth.currentUser?.uid,
        displayName: auth.currentUser?.displayName,
        timestamp: Timestamp.now(),
      },
      { merge: true },
    ).then((docRef) => {
      callback(token.key);
    });
  } catch (e) {
    console.error("Error adding token: ", e);
  }

  return;
}

export async function getToken() {
  const token = (await Notifications.getExpoPushTokenAsync()).data;
  console.log("UUUUUU:", token);
  return token;
}
