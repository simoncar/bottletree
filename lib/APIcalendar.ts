import { addDoc, collection, onSnapshot, query } from "firebase/firestore";
import { db } from "./firebase";
import { IProject, ICalendarEvent } from "./types";

type projectsRead = (projects: IProject[]) => void;

export async function getItems(projectId: string, callback: itemsRead) {
    const q = query(collection(db, "calendar"));

    const unsubscribe = onSnapshot(q, (querySnapshot) => {
        const calendarEvents: ICalendarEvent[] = [];
        querySnapshot.forEach((doc) => {
            calendarEvents.push({
                key: doc.id,
                allDay: doc.data().allDay,
                dateBegin: doc.data().dateBegin,
                dateEnd: doc.data().dateEnd,
                description: doc.data().description,
                projectId: doc.data().projectId,
                title: doc.data().title,
                uid: doc.data().uid,
                dateBeginSplit: doc
                    .data()
                    .dateBegin.toDate()
                    .toISOString()
                    .split("T")[0],
                dateEndSplit: doc
                    .data()
                    .dateEnd.toDate()
                    .toISOString()
                    .split("T")[0],
                timeBegin: doc
                    .data()
                    .dateBegin.toDate()
                    .toLocaleTimeString("en-US"),
                timeEnd: doc
                    .data()
                    .dateEnd.toDate()
                    .toLocaleTimeString("en-US"),
            });
        });

        const expandedDates = expandDateRanges(calendarEvents);
        console.log("expandedDates", expandedDates);

        callback(expandedDates);
    });

    return () => unsubscribe();
}

interface DateRange {
    dateBegin: string;
    dateEnd: string;
}

interface ExpandedDates {
    [date: string]: string;
}

function expandDateRanges(dateRanges: DateRange[]): ExpandedDates {
    const expandedDates: ExpandedDates = {};

    for (const range of dateRanges) {
        const startDate = new Date(range.dateBeginSplit);
        const endDate = new Date(range.dateEndSplit);

        const currentDate = new Date(startDate);
        while (currentDate <= endDate) {
            const dateString = currentDate.toISOString().split("T")[0];
            expandedDates[dateString] = [
                ...(expandedDates[dateString] || []),
                {
                    title: range.title,
                    description: range.description,
                    projectId: range.projectId,
                    key: range.key,
                    dateBegin: range.dateBegin,
                    dateEnd: range.dateEnd,
                    allDay: range.allDay,
                    uid: range.uid,
                    dateBeginSplit: range.dateBeginSplit,
                    timeBegin: range.timeBegin,
                    timeEnd: range.timeEnd,
                },
            ];

            currentDate.setDate(currentDate.getDate() + 1);
        }
    }

    return expandedDates;
}

export function addCalendarEvent(
    calendarEvent: ICalendarEvent,
    callback: { (id: string): void; (arg0: string): void },
) {
    try {
        addDoc(collection(db, "calendar"), {
            allDay: calendarEvent.allDay,
            dateBegin: calendarEvent.dateBegin,
            dateEnd: calendarEvent.dateEnd,
            description: calendarEvent.description,
            projectId: calendarEvent.projectId,
            title: calendarEvent.title,
            uid: calendarEvent.uid,
        }).then((docRef) => {
            callback(docRef.id);
        });
    } catch (e) {
        console.error("Error adding calendar event: ", e);
    }

    return;
}
