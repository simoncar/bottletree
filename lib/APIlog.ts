import * as Application from "expo-application";
import * as Device from "expo-device";
import { dbm } from "./firebase";
import { ILog } from "./types";

import {
  addDoc,
  collection,
  getDocs,
  orderBy,
  query,
  serverTimestamp,
} from "@react-native-firebase/firestore";

// Update getLogs to use modular API
export async function getLogs(callback: logsRead) {
  const logs: ILog[] = [];
  const logsQuery = query(
    collection(dbm, "logs"),
    orderBy("timestamp", "desc"),
  );
  const logSnapshot = await getDocs(logsQuery);

  logSnapshot.forEach((doc) => {
    const data = doc.data();
    logs.push({
      key: doc.id,
      timestamp: data.timestamp,
      loglevel: data.loglevel,
      message: data.message,
      user: data.user,
      device: data.device,
      version: data.version,
      email: data.email,
    });
  });

  callback(logs);
}

// Update addLog to use modular API
export async function addLog(log: ILog) {
  log.timestamp = serverTimestamp();
  log.device =
    Device.manufacturer +
    " > " +
    Device.modelName +
    " > " +
    Device.osName +
    " > " +
    Device.osVersion;
  log.version = Application.nativeApplicationVersion;

  try {
    const docRef = await addDoc(collection(dbm, "logs"), log);
    console.log("Log written with ID: ", docRef.id);
  } catch (e) {
    console.error("Error adding log: ", e);
  }
}
