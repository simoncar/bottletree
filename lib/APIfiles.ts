import { dbm } from "@/lib/firebase"; // or "@/lib/firebase.web"
import {
  addDoc,
  collection,
  doc,
  onSnapshot,
  orderBy,
  query,
  serverTimestamp,
} from "@react-native-firebase/firestore";

// import {
//   getDownloadURL,
//   getStorage,
//   ref as storageRef,
// } from "@react-native-firebase/storage";
import storage, { getDownloadURL } from "@react-native-firebase/storage";

import * as Crypto from "expo-crypto";
import * as DocumentPicker from "expo-document-picker";
import { setPostFile } from "./APIpost";
import { IFile, IPost, IUser } from "./types";

type FileData = {
  mimeType: string;
  name: string;
  size: number;
  uri: string;
};

type FilesJson = {
  assets: FileData[];
  canceled: boolean;
};

export async function deleteFile(
  projectId: string,
  key: string,
  callback: (id: string) => void,
) {
  try {
    const docRef = doc(
      collection(doc(collection(dbm, "projects"), projectId), "files"),
      key,
    );

    await dbm.runTransaction(async (transaction) => {
      transaction.delete(docRef);
    });

    callback(key);
  } catch (e) {
    console.error("Error deleting file from project: ", e);
  }
}
export async function getFiles(project: string, callback: filesRead) {
  const filesRef = collection(
    doc(collection(dbm, "projects"), project),
    "files",
  );
  const q = query(filesRef, orderBy("modified", "desc"));

  const unsubscribe = onSnapshot(q, (querySnapshot) => {
    const files: IFile[] = [];

    querySnapshot.forEach((docSnap) => {
      const data = docSnap.data();
      files.push({
        key: docSnap.id,
        filename: data.filename,
        url: data.url,
        mimeType: data.mimeType,
        bytes: data.bytes,
        created: data.created,
        modified: data.modified,
        projectId: data.projectId,
      });
    });

    callback(files);
  });

  return unsubscribe;
}

export async function uploadFilesAndCreateEntries(
  filesJson: DocumentPicker.DocumentPickerResult,
  projectId: string,
  user: IUser,
) {
  if (filesJson.canceled) {
    console.log("File selection was canceled.");
    return;
  }

  const projectFilesRef = collection(doc(dbm, "projects", projectId), "files");
  console.log("uploadFilesAndCreateEntries - modular:", filesJson);

  for (const file of filesJson.assets) {
    try {
      const UUID = Crypto.randomUUID();
      //const blob = await uriToBlob(file.uri);

      const filePath = `project/${projectId}/files/${UUID}`;
      const fileStorageRef = storage().ref(filePath);

      try {
        console.log("AAA111- fileStorageRef.putFile");
        await fileStorageRef.putFile(file.uri);
        console.log("AAA222- fileStorageRef.putFile");
        //await uploadBytes(fileStorageRef, blob);
      } catch (error: any) {
        if (error.code === "storage/unauthorized") {
          console.error("User does not have permission to access the object");
        } else {
          throw error;
        }
      }

      const downloadURL = await getDownloadURL(fileStorageRef);

      const fileEntry: IFile = {
        key: UUID,
        filename: file.name,
        url: downloadURL,
        mimeType: file.mimeType,
        bytes: file.size,
        created: serverTimestamp(),
        modified: serverTimestamp(),
        projectId,
      };

      await addDoc(projectFilesRef, fileEntry);

      const post: IPost = {
        key: UUID,
        caption: `${file.name}`,
        projectId,
        author: user.displayName,
        uid: user.uid,
        timestamp: serverTimestamp(),
        ratio: 1,
        file: UUID,
      };

      setPostFile(post, (postId) => {
        console.log(`Post created with ID: ${postId}`);
      });

      console.log(`File ${file.name} uploaded and entry created in Firestore.`);
    } catch (error) {
      console.error(`Error uploading file ${file.name}:`, error);
    }
  }
}
