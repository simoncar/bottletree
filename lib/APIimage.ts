import * as ImagePicker from "expo-image-picker";
import { firebase, uploadBytes } from "@/lib/firebase";
import * as Crypto from "expo-crypto";
import { Platform } from "react-native";
import * as ImageManipulator from "expo-image-manipulator";

// Types and interfaces
interface ImageAsset {
  uri: string;
  height: number;
  width: number;
}

interface ImageProcessOptions {
  compress: number;
  format: ImageManipulator.SaveFormat;
  base64: boolean;
}

interface ProgressCallback {
  (progress: number): void;
}

// Constants
const DEFAULT_COMPRESSION_RATIO = 0.2;
const IMAGE_FORMAT = ImageManipulator.SaveFormat.JPEG;
const IS_WEB_PLATFORM = Platform.OS === "web";

export const addImageFromCameraRoll = async (
  multiple: boolean,
  folder: string,
  project: string,
  progressCallback: {
    (progress: any): void;
  },
  completedCallback: {
    (sourceDownloadURLarray: any): void;
    (arg0: unknown[]): void;
  },
) => {
  const result = await ImagePicker.launchImageLibraryAsync({
    mediaTypes: ImagePicker.MediaTypeOptions.Images,
    quality: 1,
    allowsMultipleSelection: multiple,
  });

  if (!result.canceled) {
    try {
      const promises = result.assets.map((item) =>
        processItemAsync(folder, project, item, progressCallback),
      );
      const processedResults = await Promise.all(promises);
      completedCallback(processedResults);
    } catch (error) {
      console.error("Error occurred during processing:", error);
    }
  }
};

export const addImageFromPhoto = async (
  photo: any,
  folder: string,
  project: string,
  progressCallback: (progress: any) => void,
  completedCallback: { (sourceDownloadURL: any): void; (arg0: unknown): void },
) => {
  try {
    const processedResults = await processItemAsync(
      folder,
      project,
      photo,
      progressCallback,
    );
    completedCallback(processedResults);
  } catch (error) {
    console.error("Error occurred during processing addImageFromPhoto:", error);
  }
};

async function processItemAsync(
  folder: string,
  project: string,
  asset: ImageAsset,
  progressCallback: ProgressCallback,
): Promise<string> {
  let imageUri = asset.uri;
  let compressedImage;

  try {
    console.log("processItemAsync: ", asset);
    console.log("processItemAsync: ", asset.uri);

    if (IS_WEB_PLATFORM) {
      // TODO: for some reason this is not working on web, results in a larger file size and PNG?
      compressedImage = asset;
    } else {
      const compressionOptions: ImageProcessOptions = {
        compress: DEFAULT_COMPRESSION_RATIO,
        format: IMAGE_FORMAT,
        base64: true,
      };

      compressedImage = await ImageManipulator.manipulateAsync(
        asset.uri,
        [], // No operations, just compression
        compressionOptions,
      );
      imageUri = compressedImage.uri;
    }
  } catch (error) {
    console.error("Error compressing image:", error);
    throw error;
  }

  const getBlobFromUri = async (uri: string): Promise<Blob> => {
    return new Promise((resolve, reject) => {
      const xhr = new XMLHttpRequest();
      xhr.onload = () => resolve(xhr.response);
      xhr.onerror = (e) =>
        reject(new TypeError(`Network request failed: ${e}`));
      xhr.responseType = "blob";
      xhr.open("GET", uri, true);
      xhr.send(null);
    });
  };

  return new Promise((resolve, reject) => {
    try {
      const storageRef = getStorageRef(folder, project);

      if (IS_WEB_PLATFORM) {
        getBlobFromUri(imageUri).then((blob) => {
          firebase
            .storage()
            .ref(storageRef.fullPath)
            .put(blob)
            .then(() => {
              storageRef.getDownloadURL().then((downloadURL) => {
                console.log("File available at", downloadURL);
                console.log("dimensions", asset.height, asset.width);
                const ratio = asset.height / asset.width;
                resolve(ratio + "*" + downloadURL);
              });
            });
        });
      } else {
        const uploadTask = storageRef.putFile(imageUri);

        uploadTask.on(
          "state_changed",
          (snapshot) => {
            const uploadProgress = Math.floor(
              (snapshot.bytesTransferred / snapshot.totalBytes) * 100,
            );
            progressCallback(uploadProgress);

            switch (snapshot.state) {
              case "paused":
                console.log("Upload is paused");
                break;
              case "running":
                console.log(
                  "Upload is running:",
                  snapshot.bytesTransferred,
                  snapshot.totalBytes,
                );
                break;
            }
          },
          (error) => {
            console.log("error file upload:", error);
          },
          () => {
            storageRef.getDownloadURL().then((downloadURL) => {
              console.log("File available at", downloadURL);
              console.log("dimensions", asset.height, asset.width);
              const ratio = asset.height / asset.width;
              resolve(ratio + "*" + downloadURL);
            });
          },
        );
      }
    } catch (error) {
      console.log("processItemAsync error:", error);
      reject(error);
    }
  });
}

function getStorageRef(folder: string, project: string) {
  const UUID = Crypto.randomUUID();
  const d = new Date();

  const fileName = folder + "/" + project + "/images/" + UUID;

  //const storageRef = ref(storage, fileName);
  const storageRef = firebase.storage().ref(fileName);
  console.log("storageRef", storageRef);

  return storageRef;
}
