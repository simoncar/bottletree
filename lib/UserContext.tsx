import React, { createContext, useEffect, useState } from "react";
import { getUser } from "@/lib/APIuser";
import { useSession } from "@/lib/ctx";
import { IUser } from "@/lib/types";

interface UserContextType {
  user: IUser | null;
  setUser: (user: IUser) => void;
}

const UserContext = createContext<UserContextType>({
  user: null,
  setUser: () => null,
});

const UserProvider = (props: React.PropsWithChildren) => {
  const { session } = useSession();
  const [user, setUser] = useState(null);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const userData = await getUser(session);

        setUser({ ...userData });
      } catch (error) {
        console.error("error fetching user data in UserProvider", error);
      }
    };

    fetchData();
  }, []);

  return (
    <UserContext.Provider value={{ user, setUser }}>
      {props.children}
    </UserContext.Provider>
  );
};

export { UserContext, UserProvider };
