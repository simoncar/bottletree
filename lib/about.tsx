import React, { useContext, useState, useEffect } from "react";
import { Button, Pressable, StyleSheet } from "react-native";
import { Text, View } from "@/components/Themed";
import { useSession } from "@/lib/ctx";
import { auth } from "@/lib/firebase";
import { UserContext } from "@/lib/UserContext";
import { Link } from "expo-router";
import * as Application from "expo-application";
import * as Sentry from "@sentry/react-native";
import { useRouter } from "expo-router";

export const About = () => {
  const { user } = useContext(UserContext);
  const { session } = useSession();
  const [authUser, setAuthUser] = useState<string>(null);

  useEffect(() => {
    setAuthUser(auth().currentUser?.uid);
  }, []);

  const router = useRouter();
  let pressCount = 0;

  const handlePress = () => {
    pressCount += 1;
    console.log(pressCount);

    if (pressCount === 3) {
      router.push("/p");
    }
  };

  return (
    <View style={styles.aboutContainer}>
      <Pressable onPress={handlePress}>
        <Text style={styles.version}>One Build</Text>
      </Pressable>

      <Text style={styles.version}>
        {Application.nativeApplicationVersion} ({Application.nativeBuildVersion}
        ) | 12.app.sentry
      </Text>
      <Text style={styles.version}>Email: {user?.email}</Text>
      <Text style={styles.version}>Name: {user?.displayName}</Text>
      <Text style={styles.version}>User Context - {user?.uid}</Text>
      <Text style={styles.version}>Session - {session}</Text>
      <Text style={styles.version}>Auth - {auth().currentUser?.uid}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  aboutContainer: {
    alignItems: "center",
    justifyContent: "center",
    paddingBottom: 50,
    paddingTop: 50,
  },
  version: {
    color: "grey",
    fontSize: 10,
  },
});
