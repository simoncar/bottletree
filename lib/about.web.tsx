import React, { useState, useEffect, useContext } from "react";
import { StyleSheet } from "react-native";
import { getLocales } from "expo-localization";
import { Text, View } from "@/components/Themed";
import * as Application from "expo-application";
import { useSession } from "@/lib/ctx";
import { auth } from "@/lib/firebase";
import { UserContext } from "@/lib/UserContext";

export const About = () => {
  const { user } = useContext(UserContext);
  const { session } = useSession();
  const [authUser, setAuthUser] = useState<string>(true);
  const deviceLanguage = getLocales()[0].languageCode;

  useEffect(() => {
    setAuthUser(auth().currentUser?.uid);
  }, []);

  if (null == user) {
    return <View></View>;
  } else {
    return (
      <View style={styles.aboutContainer}>
        <Text style={styles.version}>{user.displayName}</Text>
        <Text style={styles.version}>{user.email}</Text>
        <Text style={styles.version}>
          Anonymous: {auth().currentUser?.isAnonymous}
        </Text>
        <Text style={styles.version}>
          {Application.nativeApplicationVersion} (
          {Application.nativeBuildVersion}) | 2
        </Text>

        <Text style={styles.version}>User Context - {user.uid}</Text>
        <Text style={styles.version}>CTX - {session}</Text>
        <Text style={styles.version}>Auth2 - {authUser}</Text>
        <Text style={styles.version}>Language - {deviceLanguage}</Text>
        <Text style={styles.version}>Project - {user.project}</Text>
      </View>
    );
  }
};

const styles = StyleSheet.create({
  aboutContainer: {
    alignItems: "center",
    justifyContent: "center",
    paddingTop: 50,
  },
  version: {
    color: "grey",
    fontSize: 14,
  },
});
