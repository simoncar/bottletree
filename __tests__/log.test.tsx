import React from "react";
import Log from "../app/(app)/log";
import { renderRouter, screen } from "expo-router/testing-library";
import { UserContext } from "@/lib/UserContext";
import { getLogs } from "@/lib/APILog";

jest.mock("@/lib/APILog", () => ({
  getLogs: jest.fn(),
}));

describe("LogList", () => {
  const user = {
    uid: "1234567890",
    displayName: "John Doe",
    email: "john.doe@example.com",
    photoURL: "https://example.com/john-doe.jpg",
    project: "73JwAXeOEhLXUggpVKK9",
  };

  const mockLogs = [
    {
      key: 1,
      device: "Android",
      logLevel: "INFO",
      message: "Some Message",
      timestamp: null,
      version: "1",
    },
    {
      key: 2,
      device: "Android",
      logLevel: "INFO",
      message: "Some Message",
      timestamp: null,
      version: "1",
    },
    {
      key: 3,
      device: "Android",
      message: "Some Message",
      timestamp: null,
      version: "1",
    },
    {
      key: 4,
      device: "Android",
      timestamp: null,
      version: "1",
    },
    {
      key: 5,
      device: "Android",
      timestamp: null,
    },
    {
      key: 6,
      timestamp: null,
      version: "1",
    },
  ];

  beforeEach(() => {
    (getLogs as jest.Mock).mockImplementation((callback) => {
      callback(mockLogs);
    });
  });

  it("render log test", async () => {
    const setUser = jest.fn();

    const MockComponent = jest.fn(() => (
      <UserContext.Provider value={{ user, setUser }}>
        <Log />
      </UserContext.Provider>
    ));

    renderRouter(
      {
        index: MockComponent,
        "note?project=73JwAXeOEhLXUggpVKK9": MockComponent,
        "(group)/b": MockComponent,
      },
      {
        initialUrl: "/",
      },
    );

    expect(screen).toHavePathname("/");
    expect(screen).toMatchSnapshot();
  });
});
