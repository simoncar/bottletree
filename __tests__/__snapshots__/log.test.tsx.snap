// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogList render log test 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNCSafeAreaView
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "flex": 1,
          "height": 200,
        }
      }
    >
      <RCTScrollView
        style={
          {
            "paddingBottom": 50,
          }
        }
      >
        <View>
          <View>
            <View>
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "borderBottomColor": "#CED0CE",
                      "borderBottomWidth": 0.5,
                      "flexDirection": "row",
                      "padding": 8,
                      "paddingVertical": 8,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "marginRight": 12,
                      "width": 50,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "borderRadius": 24,
                        "height": 48,
                        "width": 48,
                      }
                    }
                  >
                    <Text />
                  </View>
                </View>
                <View>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "fontSize": 18,
                          "marginBottom": 5,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                     
                    Some Message
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                    1
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                    Android
                  </Text>
                </View>
              </View>
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "borderBottomColor": "#CED0CE",
                      "borderBottomWidth": 0.5,
                      "flexDirection": "row",
                      "padding": 8,
                      "paddingVertical": 8,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "marginRight": 12,
                      "width": 50,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "borderRadius": 24,
                        "height": 48,
                        "width": 48,
                      }
                    }
                  >
                    <Text />
                  </View>
                </View>
                <View>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "fontSize": 18,
                          "marginBottom": 5,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                     
                    Some Message
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                    1
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                    Android
                  </Text>
                </View>
              </View>
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "borderBottomColor": "#CED0CE",
                      "borderBottomWidth": 0.5,
                      "flexDirection": "row",
                      "padding": 8,
                      "paddingVertical": 8,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "marginRight": 12,
                      "width": 50,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "borderRadius": 24,
                        "height": 48,
                        "width": 48,
                      }
                    }
                  >
                    <Text />
                  </View>
                </View>
                <View>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "fontSize": 18,
                          "marginBottom": 5,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                     
                    Some Message
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                    1
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                    Android
                  </Text>
                </View>
              </View>
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "borderBottomColor": "#CED0CE",
                      "borderBottomWidth": 0.5,
                      "flexDirection": "row",
                      "padding": 8,
                      "paddingVertical": 8,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "marginRight": 12,
                      "width": 50,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "borderRadius": 24,
                        "height": 48,
                        "width": 48,
                      }
                    }
                  >
                    <Text />
                  </View>
                </View>
                <View>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "fontSize": 18,
                          "marginBottom": 5,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                     
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                    1
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                    Android
                  </Text>
                </View>
              </View>
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "borderBottomColor": "#CED0CE",
                      "borderBottomWidth": 0.5,
                      "flexDirection": "row",
                      "padding": 8,
                      "paddingVertical": 8,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "marginRight": 12,
                      "width": 50,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "borderRadius": 24,
                        "height": 48,
                        "width": 48,
                      }
                    }
                  >
                    <Text />
                  </View>
                </View>
                <View>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "fontSize": 18,
                          "marginBottom": 5,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                     
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                    Android
                  </Text>
                </View>
              </View>
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "borderBottomColor": "#CED0CE",
                      "borderBottomWidth": 0.5,
                      "flexDirection": "row",
                      "padding": 8,
                      "paddingVertical": 8,
                    },
                    {
                      "backgroundColor": "transparent",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "marginRight": 12,
                      "width": 50,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "borderRadius": 24,
                        "height": 48,
                        "width": 48,
                      }
                    }
                  >
                    <Text />
                  </View>
                </View>
                <View>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "fontSize": 18,
                          "marginBottom": 5,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                     
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  />
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  >
                    1
                  </Text>
                  <Text
                    style={
                      [
                        {
                          "color": "#000",
                        },
                        {
                          "color": "grey",
                          "fontSize": 14,
                        },
                      ]
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </RNCSafeAreaView>
</RNCSafeAreaProvider>
`;
